In [ ]:
#Previo a la ejecución del taller, es imperativo instalar las librerias que se usaran para el desarrollo del mismo 
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.optimize import minimize
from scipy.optimize import linprog
import cvxpy as cp

1) La empresa XYZ produce dos productos: A y B. La ganancia por unidad de A es de 3 dólares y de B es de 5 dólares. Para su producción dispone de 500 unidades de materia prima, 1000 unidades de mano de obra y 2000 dólares de capital, de los cuales se necesitan para producir una unidad de:
###### i) A: una unidad de materia prima, dos unidades de mano de obra y una unidad de capital.
###### ii) B: una unidad de materia prima, cuatro unidades de mano de obra y dos unidades de capital. Además, se necesita producir al menos 100 unidades de A y 150 unidades de B para satisfacer la demanda del mercado.

Además, se necesita producir al menos 100 unidades de A y 150 unidades de B para
satisfacer la demanda del mercado.

Encuentre la asignación óptima de A y B que maximizan las ganancias de la empresa. Para ello, formule un problema de LP utilizando Scipy.

In [ ]:
# Función objetivo
c1 = np.array([-3, -5])
# Restricciones
A1 = np.array([[1, 1], [2, 4], [1, 2]])
b1 = np.array([500, 1000, 2000])
bounds1 = ((100, None), 
          (150, None))
# Optimización
R1 = linprog(c1, A_ub=A1, b_ub=b1, bounds=bounds1)
pd.DataFrame(R1.x).T


0      1
0  200.0  150.0

In [ ]:
# Ganancia total
-R1.fun

1350.0

Segun los datos que nos dan y con las formulacioines que se hicieron, podemos determinar que la asignacion optima de A y B que maximizan las ganancias de la empresa don de 200 y 150 respectivamente. Con con maximizacion de ganancias total de 1350

2) Usted desea invertir la suma de $100 millones en cuatro fondos que tienen diferentes
retornos esperados (E(R)), así como diferentes ponderaciones en acciones de alto valor,
gran tamaño y alta liquidez, como muestra la siguiente tabla: 

![imagen especial](https://drive.google.com/uc?id=1gzaw-FoCup08aCbr--ZLRqZL-k11AHUg) 

La asignación anterior debe contener al menos un 35% de acciones de alto valor, un 30%
en tamaño y un 15% de acciones de liquidez. Encuentre una asignación factible con el
retorno esperado más alto, bajo el supuesto de que solo se le permite mantener posiciones
largas en los fondos.

Formule y resuelva el problema de LP considerando las siguientes restricciones:

![imagen especial](https://drive.google.com/uc?id=1i3CaJ62ODAHU1LWYWg0OfXF1MDUnf7MY)

Nota: resuelva el problema anterior usando las librerías de Scipy y CVXPY de Python.


In [ ]:
# Parametros de la función objetivo
c = np.array([0.10,0.15,0.16,0.08])
# Restricciones de desigualdad
A_ub = np.array([[0.5,0.3,0.25,0.6],
                  [0.3,0.1,0.4,0.2],
              [0.2,0.6,0.35,0.2]])

b_ub = np.array([0.35,0.30,0.15])
A_eq=np.array([[1,1,1,1]])
b_eq=np.array([1])

bounds_=  [(  0,    None),
              (0, None),
              (0, None),
              (  0,   None)]

# Solución del problema
sol = linprog(-c, A_ub=-A_ub, b_ub=-b_ub,A_eq=A_eq,b_eq=b_eq, bounds=bounds_,method='simplex')
pd.DataFrame(sol.x).round(4).T


<ipython-input-2-155117053682>:18: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  sol = linprog(-c, A_ub=-A_ub, b_ub=-b_ub,A_eq=A_eq,b_eq=b_eq, bounds=bounds_,method='simplex')


0       1       2       3
0  0.0  0.1579  0.5789  0.2632

In [ ]:
#Función Objetivo (a Maximizar)

0.13736842105263158

In [ ]:
# Ejecución con Librería CVXPY
n = 4
x = cp.Variable(n)
objetivo = cp.Minimize(-c.T@x)
restricciones = [A_ub @ x >= b_ub, cp.sum(x) == 1, x >= 0]
prob = cp.Problem(objetivo, restricciones)
prob.solve()
sol_2= x.value
print(round(pd.DataFrame(sol_2).T,4))

     0       1       2       3
0  0.0  0.1579  0.5789  0.2632


Utilizando los dos metodos podemos llegar al mismo resultado, donde vemos que la asignacion mas factible bajos los parametros anteriormente mencionados son: 0.0 en el fondo 1, 0.1579 en el fondo 2, 0.5789 en el fondo 3 y 0.2632 en el fondo 4. Teniendo en cuenta aspectos como el tamaño, el valor y la liquidez 

3) La empresa XYZ se enfrente al siguiente problema de manejo de efectivo. Las cifras
están expresadas en millones ($).

![imagen especial](https://drive.google.com/uc?id=1_kJDWi3NTrM8x2D_Q8cOwC-ux0CcxbbS)

Además, se sabe que la empresa cuenta con las siguientes fuentes de recursos:

* Una línea de crédito de hasta $100 a una tasa de interés del 1% mensual.
* Puede emitir papel comercial a 90 días con un interés total del 2% por un período
de 3 meses.
* Todos los meses se pueden invertir los fondos sobrantes a una tasa de interés del
0,3% mensual. Por ejemplo, la empresa podría hacer un depósito bancario que le
reconoce esa tasa.

Sobre este problema, la empresa desea saber si: ¿usar la línea de crédito en algunos de
los meses le representa un alto costo? si es así ¿cuándo debería usarla? ¿cuánto cuesta?
¿qué pagos de intereses tendrá que hacer entre enero y junio? etc... Para responder a
estos interrogantes, la empresa puede formular y resolver un problema de LP.

Para la
formulación del problema tener en cuenta lo siguiente:



![imagen especial](https://drive.google.com/uc?id=16z_fHH87f89oti3G55cGwkzRH0H7qKT7)

Nota: resuelva el problema anterior usando las librerías de Scipy y CVXPY de Python.

In [ ]:
# Libreria Scipy
c3 = np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1])
r_13 = np.array([[1, 0, 0, 0, 0, 1, 0, 0, -1, 0, 0, 0, 0, 0],
                    [-1.01, 1, 0, 0, 0, 0, 1, 0, 1.003, -1, 0, 0, 0, 0],
                    [0, -1.01, 1, 0, 0, 0, 0, 1, 0, 1.003, -1, 0, 0, 0],
                    [0, 0, -1.01, 1, 0, -1.02, 0, 0, 0, 0, 1.003, -1, 0, 0],
                    [0, 0, 0, -1.01, 1 ,0, -1.02, 0, 0, 0, 0, 1.003, -1, 0],
                    [0, 0, 0, 0, -1.01, 0, 0, -1.02, 0, 0, 0, 0, 1.003, -1]])

r_23 = np.array([150, 100, -200, 200, -50, -300])

R3 = linprog(-c3, A_eq = r_13, b_eq = r_23, method = 'simplex')
pd.DataFrame(R3.x).round(2).T

<ipython-input-14-a214ef56d6bc>:12: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  R3 = linprog(-c3, A_eq = r_13, b_eq = r_23, method = 'simplex')


0      1    2    3    4      5      6       7    8    9       10   11  \
0  0.0  50.98  0.0  0.0  0.0  150.0  49.02  203.43  0.0  0.0  351.94  0.0   

    12    13  
0  0.0  92.5

In [ ]:
# Librería CVXPY
n = 14
x = cp.Variable(n)
o3 = cp.Minimize(-c3.T@x)
r33 = [r_13 @ x == r_23, x >= 0]
prob = cp.Problem(o3, r33)
prob.solve()
r3_2= x.value
print(round(pd.DataFrame(r3_2).T,2))

    0      1    2    3     4      5      6       7    8    9       10   11  \
0  0.0  42.69  0.0  0.0  8.46  150.0  57.31  195.06  0.0  0.0  351.94  0.0   

    12    13  
0  0.0  92.5  


Inicialmente debemos tener en cuneta que en este caso no van a dar 100% igual con los dos metodos, debido a la manera en que optimizamos una libreria respecto a la otra. Si es importante usar el credito, pero es evidente que le genera alto costos, mayormente del septimo mes en adelatnte, ya que del sexto mes hacia atras solo se vera afectado un poco mas de 250.

4) La empresa XYZ desea encontrar el saldo óptimo que debe mantener en su cuenta bancaria en un horizonte de un mes buscando maximizar su rentabilidad total. Se sabe que la empresa espera recibir entradas de efectivo de: 500 dólares en la primera semana, 300 dólares en la segunda semana, 200 dólares en la tercera semana y $400 en la cuarta semana. Todas las cifras se encuentran en millones de pesos. Además, la empresa debe realizar los siguientes pagos durante el mes: 350 dólares en la primera semana, 400 dólares en la segunda semana, 250 dólares en la tercera semana y 200 dólares en la cuarta semana y, para ello, debe mantener un saldo suficiente que respalde estas obligaciones.

Para aprovechar la rentabilidad que ofrecen los saldos en cada momento, la empresa puede invertir sus recursos en un instrumento a corto plazo en el que recibe un rendimiento del 0,02% por día, pero al realizar esta operación debe incurrir en un costo de transacción del 0,01% del monto transferido.

Formule un problema de LP utilizando Scipy que le permita maximizar la rentabilidad total de los saldos en su cuenta teniendo en cuenta las restricciones de sus flujos de efectivo.

In [ ]:
# Función objetivo
c4 = np.array([0, 0, 0, 0, 0, 0, 0, 1])
# Restricciones
r4 = (1+0.0001)**7
A4 = np.array([[1, 0, 0, 0, -1, 0, 0, 0], 
                [0, 1, 0, 0, r4, -1, 0, 0], 
                [0, 0, 1, 0, 0, r4, -1, 0], 
                [0, 0, 0, 1, 0, 0, r4, -1]])
b4 = np.array([[150, -100, -50, 200]])
# Función de optimización
R4 = linprog(-c4, A_ub=A4, b_ub=-b4, method='simplex')
RTA4 = pd.DataFrame(R4.x).T
RTA4.round(2)

<ipython-input-3-01222c0f90b9>:11: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  R4 = linprog(-c4, A_ub=A4, b_ub=-b4, method='simplex')


0    1    2    3      4      5     6       7
0  0.0  0.0  0.0  0.0  150.0  50.11  0.14  200.14

In [ ]:
#Valor Máximo (Objetivo)
-R4.fun

200.14021366142958

Teniendo en cuenta un movimiento de 7 semanas analizando tanto ingresos como egresos podemos observar que el saldo optimo que se debe tener en la cuenta es de 200.1402

5) La empresa XYZ tiene que decidir como distribuye su presupuesto $10.000 millones y la mano de obra (horas de trabajo) en tres departamentos: ventas, marketing y producción buscando maximizar sus ganancias. Para ello, se sabe que cada departamento le proporciona a la empresa las siguientes ganancias: ventas: 5 dólares, marketing: 8 dólares, y producción: 10 dólares. Esta asignación debe cumplir las siguientes restricciones:

i. El total de horas de trabajo disponibles es de 1.000.

ii. El departamento de ventas debe recibir al menos el 20% del presupuesto total.

iii. El departamento de marketing debe recibir al menos el 30% del total de horas de trabajo disponibles.

iv. El departamento de producción debe recibir al menos el 25% del total de horas de trabajo disponibles.

Formule un problema de LP utilizando Scipy que le permita maximizar

In [ ]:
# Función objetivo
c5 = np.array([0, 0, 0, 5, 8, 10])
# Restricciones
A5 = np.array([[0, 0, 0, 1, 0, 0],
                [0, 1, 0, 0, 0, 0],
                [0, 0, 1, 0, 0, 0],
                [0, 0, 0, 0, 1, 0],
                [0, 0, 0, 0, 0, 1]])
b5 = np.array([2000, 300, 250, 2000, 2000])
A_5_ = np.array([[1, 1, 1, 0, 0, 0],
                  [0, 0, 0, 1, 1, 1]])
b_5_= np.array([1000, 10000])
# RT
R5 = linprog(-c5, A_ub = -A5, b_ub = -b5, A_eq = A_5_, b_eq = b_5_, method = 'simplex')
pd.DataFrame(R5.x).round(4).T


<ipython-input-18-fc0d4fd4b667>:14: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  R5 = linprog(-c5, A_ub = -A5, b_ub = -b5, A_eq = A_5_, b_eq = b_5_, method = 'simplex')


0      1      2       3       4       5
0  450.0  300.0  250.0  2000.0  2000.0  6000.0

In [ ]:
-R5.fun

86000.0

Teniendo en cuenta los anteriores datos podemos deducir que las mejor forma de distribuir el presupuesto es la siguiente manera: 2000 para ventas, 2000 para marketing y 6000 para produccion, teniendo una ganancia total de 86000

6) Considere una empresa con tres puntos potenciales para ubicar sus instalaciones (J =
{1,2,3}) y cinco puntos de demanda que debe satisfacer (I = {1,2,3,4,5}). La apertura del
punto j ∈ J tiene un costo fijo de funcionamiento fj, independientemente del volumen de ventas que maneje. Este volumen está limitado a una determinada cantidad máxima que puede manejarse anualmente (Mj). Además, hay un costo de transporte cij para ir desde el punto j para abastecer la demanda en el punto i. Las siguientes tablas muestran los datos de demanda en cada punto:

![imagen especial](https://drive.google.com/uc?id=1WVmYwfKXEpTb6ArphVDY7SrIl5-xwqpw)

Y los costos de transporte, de funcionamiento y cantidad máxima que puede mantener en cada

![imagen especial](https://drive.google.com/uc?id=1aGidOx0gVLI_5QadccvTbgurjDX4Tox7)

Las columnas 1-5 muestran el costo de transporte desde cada punto instalado (planta) a
cada demanda que debe satisfacer. Las columnas fj y Mj muestran los costos fijos y la
capacidad máxima en cada ubicación. Encuentre la ubicación óptima para satisfacer las
demandas de tal forma que se minimicen los costos.
Formule un problema de optimización utilizando Scipy que le permita encontrar la
estrategia optima de la empresa para reducir sus costos de funcionamiento y transporte.

In [ ]:
# Función objetivo

c6 = np.array([[4, 5, 6, 8, 10],
               [6, 4, 3, 5, 8],
               [9, 7, 4, 3, 4]])

# Restricciones
A6 = np.array([500, 500, 500])
d6 = np.array([80, 270, 250, 160, 180])
x6 = cp.Variable((3, 5), nonneg=True)
matriz = np.array([1000, 1000, 1000])
cy6 = cp.Variable(3, boolean = True)
O6= cp.Minimize(cp.sum(cp.multiply(c6, x6)))
R6 = [cp.sum(x6, axis=1)<= A6 @ cy6, cp.sum(x6, axis=0)==d6, cp.sum(cy6*A6) >= cp.sum(d6)]

# Optimización
Rta6 = cp.Problem(O6, R6)
Rta6.solve()
print(round(pd.DataFrame(x6.value),4))
cy6.value

      0      1      2      3      4
0  80.0   -0.0   -0.0   -0.0   -0.0
1  -0.0  270.0  250.0   -0.0   -0.0
2  -0.0   -0.0   -0.0  160.0  180.0


/usr/local/lib/python3.9/dist-packages/cvxpy/expressions/expression.py:593: UserWarning: 
This use of ``*`` has resulted in matrix multiplication.
Using ``*`` for matrix multiplication has been deprecated since CVXPY 1.1.
    Use ``*`` for matrix-scalar and vector-scalar multiplication.
    Use ``@`` for matrix-matrix and matrix-vector multiplication.
    Use ``multiply`` for elementwise multiplication.
This code path has been hit 1 times so far.

  warnings.warn(msg, UserWarning)


array([1., 1., 0.])

Analizando y uniendo todos los datos podemos determinar: 1. la palnta numero 1 se deja cerrada. 2. con la planta 2 se va a satisacer el total del punto 1 y 2. 3.  con la planta numero 3 se va a satisacer por completo el punto numero 4 y 5. 4. Entre la planta 2 y 3 se va a satisfacer el punto numero 3, produciendo 150 y 100 respectivamente. 